### DISCUSSION TITLE
Is there an existing way to generate Entity Framework Entities and domain DTOs like there is for controller generation?

### DISCUSSION THREAD

**Discussion by DanMarshall909** [RicoSuter/NSwag#4569] [Q&A]
Hi all,

I'd like to generate EF entities and domain objects (ideally with extension methods to map them) along side request and response objects.

It seems like a logical extension to what the code generation features do already but I can't find any templates that do this.

If the codegen doesn't exist I'd be eager to do it as a personal project.

Cheers,

**Comment by DennyFiguerres** [comment]
Dan have you looked at OData ?   it can take an EF model and make webapi that are based on the ef models. OData also enables a form of queries to be made via the api so that the api server uses LINQ to query the EF model and return objects back that can include related objects.

the interop between OData and OpenApi is not great but OData has a huge ability to work with EF and SQL.

https://www.odata.org/

https://github.com/OData/RESTier

https://learn.microsoft.com/en-us/odata/webapi-8/getting-started?tabs=net60%2Cvisual-studio-2022%2Cvisual-studio

OData is used in a lot of places.....

**Reply by DanMarshall909** [reply]
Thanks so much for the reply. 

Yes, I have used it before in a limited fashion. It's great and I need to get into it more, but what I am talking about is really the reverse i.e. generating the EF models and context from an API specification.

I know this is a bit different to most work flows, but in our case we found it very useful to allow front and back end developers to collaborate on the specification at design time, and then update and regenerate controllers and models during development as required. A pull request is far easier to approve - you're just checking some yaml. The front end client gets regenerated immediately along with interfaces, so they can start mocking from that without even waiting for a real API.

It would just be convenient to be able to have the EF stuff ready to pop in place once I regenerate the API too.

**Reply by DennyFiguerres** [reply]
ok, do you get thought that with OData the developers can make the select queries that get the right data without chnaging the back end data store ?

when you refer to re-generating the EF that to me sounds like a possible problem are you chnaging the database design ? or what are your ef models based on ?
